sycl/doc/extensions/supported/sycl_ext_intel_dataflow_pipes.asciidoc

```diff
@@ -601,6 +679,38 @@ The above example function `pipe_memcpy()` could alternatively be templated on t
 
 Automated mechanisms are possible to provide uniquification across calls, and could be exposed through a wrapper or library.
 
+== Add new copy and memcpy members to the handler class
+
+Add the following functions to the `sycl::handler` interface described in Section 4.9.4.3 of
+the SYCL 2020 specification.
+
+Add to Table 130, "Member functions of the handler class".
+
+--
+[options="header"]
+|====
+| Member Function | Description
+a| 
+[source, c++]
+----
+void ext_intel_read_write_host_pipe
+       (const std::string &Name, void *Ptr,
+                       size_t Size, bool Block, bool Read);
```

Comment:
This new function seems a little unorthodox for a SYCL interface. I would have expected a separate read and write function, both returning an event you could wait on, instead of specifying whether it block.

@gmlueck - Thoughts?

Reply:
@steffenlarsen , thanks for the review. The underlying opencl function also needs to know whether this call is a blocking call or non-blocking call.

```
ACL_EXPORT
CL_API_ENTRY cl_int CL_API_CALL clEnqueueWriteHostPipeINTEL(
                                    cl_command_queue command_queue,
                                    cl_program program,
                                    const char* pipe_symbol,
                                    cl_bool blocking_write,
                                    const void* ptr,
                                    size_t size,
                                    cl_uint num_events_in_wait_list,
                                    const cl_event* event_wait_list,
                                    cl_event* event)
```

I think this was the reason why the Block was passing to the other functions.

Reply:
Other OpenCL interfaces have that too, but to my knowledge we don't have any SYCL interfaces that utilize that. For example, `piextEnqueueDeviceGlobalVariableWrite` (inspired by `clEnqueueWriteGlobalVariable` from another OpenCL extension) but we do not use it as part of the memcpy to/from device_global interfaces.

Reply:
If we believe there is a benefit to having it, I am okay with keeping the argument. As for the `Read` parameter, I still think it would be better to have separate functions, similar to `read`/`write` on `pipe`. That way we could also have `Block` be optional.

Also, would it be possible for this interface to return an event so the user can create appropriate dependencies for this?

Reply:
Hi @steffenlarsen, Thanks! Yes I can create separate functions for read and write and make `Block` optional. 

Regarding the event I am not sure what I should do.  

```
void handler::ext_intel_read_write_host_pipe(const std::string &Name, void *Ptr,
                                             size_t Size, bool Block,
                                             bool Read) {
  MImpl->HostPipeName = Name;
  MImpl->HostPipePtr = Ptr;
  MImpl->HostPipeTypeSize = Size;
  MImpl->HostPipeBlocking = Block;
  MImpl->HostPipeRead = Read;
  setType(detail::CG::ReadWriteHostPipe);
}
```
In the `commands.cpp`,  I had this code creating the event.
```
  case CG::CGTYPE::ReadWriteHostPipe: {
    CGReadWriteHostPipe *ExecReadWriteHostPipe =
        (CGReadWriteHostPipe *)MCommandGroup.get();
    std::string pipeName = ExecReadWriteHostPipe->getPipeName();
    void *hostPtr = ExecReadWriteHostPipe->getHostPtr();
    size_t typeSize = ExecReadWriteHostPipe->getTypeSize();
    bool blocking = ExecReadWriteHostPipe->isBlocking();
    bool read = ExecReadWriteHostPipe->isReadHostPipe();

    if (!Event) {
      Event = &MEvent->getHandleRef();
    }
    return enqueueReadWriteHostPipe(MQueue, pipeName, blocking, hostPtr,
                                    typeSize, RawEvents, Event, read);
  }
```

Should I move the event creation somewhere else?

Reply:
Sorry, disregard the event. In my head this was a queue shortcut, but the way it is here you get the event from the call to submit, so everything is fine in that regard.

We may want a queue shortcut in the future, but that can wait.

Reply:
@steffenlarsen , thanks! I have made the change to the doc and also made the corresponding code changes into the https://github.com/intel/llvm/pull/7468.